AX: AXCoreObject::supportsPressAction should not be true for things that are pointer-events:none

twilco opened the pull request:
#### c916ef63d6b0178617afe81bd25c5de6ba2616b6
<pre>
AX: AXCoreObject::supportsPressAction should not be true for things that are pointer-events:none
<a href="https://bugs.webkit.org/show_bug.cgi?id=299604">https://bugs.webkit.org/show_bug.cgi?id=299604</a>
<a href="https://rdar.apple.com/161409435">rdar://161409435</a>

Reviewed by Joshua Hoffman.

pointer-events:none is an explicit signal that something is not clickable or pressable, so let&apos;s respect that in
AXCoreObject::supportsPressAction.

* LayoutTests/accessibility/mac/supports-press-action-pointer-events-none-expected.txt: Added.
* LayoutTests/accessibility/mac/supports-press-action-pointer-events-none.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::supportsPressAction const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXNotifications.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onStyleChange):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::convertToPropertyFlag):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/300596@main">https://commits.webkit.org/300596@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8faeeb5c9d877d5f010c5fba0c298b61e5c7893b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123145 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42859 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33556 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129816 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75257 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a6dd4fee-0e26-4a13-8e2f-023be21de485) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125022 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43583 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51454 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93615 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126096 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34740 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110217 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74247 "3 api tests failed or timed out") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/670d2b16-f455-4917-ae60-a1c40fe18dd6) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33708 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28373 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73323 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104452 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28599 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132534 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50095 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38142 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102108 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50471 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106438 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101964 "Found 1 new API test failure: WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25927 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47337 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25541 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46868 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49950 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55711 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49419 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52770 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51099 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityNodeObject.h:
```diff
@@ -196,12 +197,10 @@ class AccessibilityNodeObject : public AccessibilityObject {
     bool hasCursorPointer() const final
```
Nit: What do you think about renaming this method to something like `hasValidCursorPointer()`? Based on the check within the body, it seems like something can get a pointer cursor, but also not accept cursor events, which feel like two distinct things (especially considering we also have a separate `hasPointerEventsNone` method).

twilco replied:
`pointer-events:none` actually causes a cursor pointer to not be visually rendered despite the presence of `cursor:pointer`, so I think it's still accurate!

hoffmanjoshua replied:
Gotcha, makes sense!

minorninth commented on LayoutTests/accessibility/mac/supports-press-action-pointer-events-none.html:
```diff
@@ -0,0 +1,44 @@
+if (window.accessibilityController) {
+    window.jsTestIsAsync = true;
+
+    var text = accessibilityController.accessibleElementById("inaccessible-button").childAtIndex(0);
```
Would isPressActionSupported also return true on the group and any other intermediate descendants if any?

twilco replied:
It would return true for the group because of [this check for whether this has a click handler](https://github.com/WebKit/WebKit/blob/752e87ed0dc7dcad7a0787defb8cb6ca175bb985/Source/WebCore/accessibility/AXCoreObject.cpp#L1408). I think this is fine since no AT I'm aware of explicitly views the group as a navigation stop, but I think we could also remove this since the valuable markup repairs that I'm aware of take place in the [static text / image branch](https://github.com/WebKit/WebKit/blob/752e87ed0dc7dcad7a0787defb8cb6ca175bb985/Source/WebCore/accessibility/AXCoreObject.cpp#L1411). This also doesn't check for cursor:pointer, just the presence of a click handler, which isn't great, so at the very least that should probably change too.

As for other descendants — it would only return true for the first [matching](https://github.com/WebKit/WebKit/blob/752e87ed0dc7dcad7a0787defb8cb6ca175bb985/Source/WebCore/accessibility/AXCoreObject.cpp#L1444) descendant, and false if there are [too many matches](https://github.com/WebKit/WebKit/blob/752e87ed0dc7dcad7a0787defb8cb6ca175bb985/Source/WebCore/accessibility/AXCoreObject.cpp#L1454), indicating a too-coarse click handler.

Only returning true for the first matching descendant ensures that users don't perceive "clickable" (in the case of VoiceOver) multiple times in a row.

minorninth reviewed (commented):
lgtm

twilco committed:
AX: AXCoreObject::supportsPressAction should not be true for things that are pointer-events:none
https://bugs.webkit.org/show_bug.cgi?id=299604
rdar://161409435

Reviewed by Joshua Hoffman.

pointer-events:none is an explicit signal that something is not clickable or pressable, so let's respect that in
AXCoreObject::supportsPressAction.

* LayoutTests/accessibility/mac/supports-press-action-pointer-events-none-expected.txt: Added.
* LayoutTests/accessibility/mac/supports-press-action-pointer-events-none.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::supportsPressAction const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXNotifications.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onStyleChange):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::convertToPropertyFlag):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/300596@main
